### PR TITLE
use weeder

### LINKED ISSUE #2043: Use `weeder` for dead code analysis
See https://github.com/swarm-game/swarm/pull/1979#discussion_r1660042482:

> It might also be a good idea to use a tool to find dead code so we can identify it sooner.

### PR DESCRIPTION
**Pull request by kostmo**
Closes #2043

# Local usage notes

I suspect that HLS in VS Code spontaneously pollutes the `.hie` directory, because after a few successful invocations, `weeder` started complaining with:
```
incompatible hie file: /home/kostmo/github/swarm/.hie/Swarm/Web/Auth.hie
    this version of weeder was compiled with GHC version 9064
    the hie files in this project were generated with GHC version 9048
    weeder must be built with the same GHC version as the project it is used on
```

## Fixing false positives

Previously, for each `library` and `executable`, the HIE directory was always the same:
```
ghc-options:
    -hiedir=.hie
```
However, most of the executables have at least one module path that is the same relative to their `hs-source-dirs` base: namely, `Main.hs`.  This resulted in all but one of the `Main.hie` files being overwritten in the end.

To avoid this, I have specified a different `-hiedir` for each `library`/`executable`/`test` that parallels its `hs-source-dirs` base.  This way, so long as `hs-source-dirs` are unique across these targets, all of the `*.hs` files across the entire project will have unique `*.hie` paths.

## Whitelisting exceptions

There are some known limitations with `weeder` regarding support for Template Haskell and type families.  Exceptions are specified in the `roots` list in `weeder.toml`.

After removing a handful of dead functions myself, there are approx. 30 "true positive" weeds that I have listed explicitly in `weeder.toml`.  Maintenance of this list of exceptions should eventually be easier with https://github.com/ocharles/weeder/issues/146.

# Integration with CI

I found a ready-made GitHub Action for weeder: https://github.com/freckle/weeder-action
I hacked support directly into the generated `.github/workflows/haskell-ci.yml` file.

Ideally, the generator would have an option for a `weeder` step.  Indeed, there is an open issue for supporting `weeder` in `haskell-ci`: https://github.com/haskell-CI/haskell-ci/issues/132

A separate but related functionality that would be nice in `haskell-ci` is to specify **one** of the GHC versions in the matrix to do additional validations/builds that would be redundant if performed on the other matrix entries.  I suppose `weeder` is inexpensive enough to redo four times, but the `weeder` binary is not available for download for all GHC versions (e.g. ghc `9.8.2`).  Something like `haddock` we probably only need to build once.
I have hacked this in to the generated file for `weeder` with a simple [`if` condition](https://github.com/swarm-game/swarm/pull/2044/files#diff-73f8a010e9b95aa9fe944c316576b12c2ec961272d428e38721112ecb1c1a98bR227).

### CODE REVIEW

**Comment by xsebek** [comment]
I tried it too, but gave up after the false positives. I did not notice that running HLS in VSCode has an impact, maybe I should give it a second try. 🤔

File: weeder.toml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+roots = [
+    "^Main.main$",
```
Comment (xsebek) [inline weeder.toml:2]: I think this should get all our executables and their dependencies, but it's probably buggy. 😕

File: .github/workflows/haskell-ci.yml [hunk 2]
```diff
@@ -203,6 +204,8 @@ jobs:
           $HCPKG list --simple-output --names-only | perl -ne 'for (split /\s+/) { print "constraints: any.$_ installed\n" unless /^(swarm)$/; }' >> cabal.project.local
           cat cabal.project
           cat cabal.project.local
+          echo "::set-output name=SWARM_PKGDIR::${PKGDIR_swarm}"
```
Comment (byorgey) [inline .github/workflows/haskell-ci.yml:207]: As @nitinprakash96 [pointed out to me](https://github.com/swarm-game/swarm/issues/1970#issuecomment-2226885910), `set-output` is deprecated.

**Review by byorgey — APPROVED** [review]
Great work, thanks!

My concern with editing `haskell-ci.yml` manually like this is that these changes might be lost next time we automatically regenerate it.  At the very least, we should add some clear comments to the top of the file explaining which parts need to be left alone (we already have such a comment, about the custom `on.push` and `on.pull-request` conditions).

**Review by xsebek — APPROVED** [review]
Amazing, thanks @kostmo! 👍

File: weeder.toml
[the remark this answers is not in the document]
Comment (kostmo) [reply weeder.toml:1]: It doesn't have to be; it just seems to be common convention.
>  I wish all our Haskell CI configurations were a bit more hidden.

Yes, this is in the esteemed company of `fourmolu.yaml`, `.stan.toml`, and `.hlint.yaml`.  Maybe someday we'll move them.

[the remark this answers is not in the document]
Comment (byorgey) [reply weeder.toml:1]: But we did [just move a bunch of other ones](https://github.com/swarm-game/swarm/pull/2037)...

[the remark this answers is not in the document]
Comment (kostmo) [reply weeder.toml:1]: Created #2057

**Pull request merged**